### PR TITLE
Less iffy makefile for flattening policy

### PR DESCRIPTION
**Pull request by oharboe**
No new or removed features, just moving conditional code into .tcl.

Introducing a do-synth-flatten-policy stage that figures out which modules to keep in flattening, with SYNTH_HIERARCHICAL=0, all Verilog modules are flattened.

### CODE REVIEW

File: flow/designs/asap7/aes-block/config.mk [hunk 1]
```diff
@@ -30,3 +30,6 @@ export PWR_NETS_VOLTAGES      =
 
 # The macros are very small so use a smaller halo
 export MACRO_PLACE_HALO        ?= 5 5
+
+export CACHED_FLATTEN_POLICY = ./designs/$(PLATFORM)/$(DESIGN_NICKNAME)/flatten.tcl
+export ADDITIONAL_FILES = $(CACHED_FLATTEN_POLICY)
```
Comment (maliberty) [inline flow/designs/asap7/aes-block/config.mk]: Why is this necessary for this one design (along with flatten.tcl)?

File: flow/Makefile [hunk 2]
```diff
@@ -541,7 +537,7 @@ $(RESULTS_DIR)/1_synth.rtlil:
 	$(UNSET_AND_MAKE) do-yosys-canonicalize
 
 $(RESULTS_DIR)/1_1_yosys.v: $(RESULTS_DIR)/1_synth.rtlil
-	$(UNSET_AND_MAKE) do-yosys
+	$(UNSET_AND_MAKE) do-yosys-flatten-policy do-yosys
```
Comment (maliberty) [inline flow/Makefile]: It looks like SYNTH_HIERARCHICAL is no longer supported.  Is that correct and if so, why?

Comment (oharboe) [reply flow/Makefile]: Still supported, but the iffy code from Makefile is gone. It is all handled in .tcl now.

File: flow/designs/asap7/aes-block/config.mk (hunk 1 above)
Comment (oharboe) [reply flow/designs/asap7/aes-block/config.mk]: aes-block is more of an ORFS script test case for blocks and other and not really a quality of results/real design, so it seems like a good choice for testing this feature.

File: flow/scripts/synth_hier_report.tcl [hunk 3]
```diff
@@ -69,3 +75,4 @@ if { [info exist ::env(REPORTS_DIR)] && [file isfile $::env(REPORTS_DIR)/synth_h
   close $out_script_ptr
 }
 
+write_flatten_policy
```
Comment (maliberty) [inline flow/scripts/synth_hier_report.tcl]: The naming feels a bit backwards throught.  The file contains the list of things _not_ to flatten.

Comment (oharboe) [reply flow/scripts/synth_hier_report.tcl]: That to me is an implementation detail.

In a compiler we call it inlining policy regardless of how it is implemented. unflattening policy implies to me that the modules dont  exist, but have to be created... but they exist and we have a policy of which to flatten.

Comment (maliberty) [reply flow/scripts/synth_hier_report.tcl]: You have promoted this to a result and not an implementation detail.  It was mark_hier_stop_modules.tcl; perhaps just keep_hierarchy.tcl

Comment (oharboe) [reply flow/scripts/synth_hier_report.tcl]: fixed.

File: flow/scripts/synth_hier_report.tcl [hunk 4]
```diff
@@ -1,22 +1,28 @@
-source $::env(SCRIPTS_DIR)/synth_preamble.tcl
+proc write_flatten_policy {} {
```
Comment (maliberty) [inline flow/scripts/synth_hier_report.tcl]: Please update to match

File: flow/Makefile [hunk 5]
```diff
@@ -508,15 +507,12 @@ memory:
 export SYNTH_SCRIPT ?= $(SCRIPTS_DIR)/synth.tcl
 export SYNTH_MEMORY_MAX_BITS ?= 4096
 
-$(SYNTH_STOP_MODULE_SCRIPT):
+.PHONY: do-yosys-flatten-policy
+do-yosys-flatten-policy:
```
Comment (maliberty) [inline flow/Makefile]: update to match

**Review by maliberty — CHANGES_REQUESTED** [review]

File: flow/scripts/synth_hier_report.tcl (hunk 4 above)
Comment (oharboe) [reply flow/scripts/synth_hier_report.tcl]: fixed

File: flow/Makefile (hunk 5 above)
Comment (oharboe) [reply flow/Makefile]: fixed

**Review by maliberty — APPROVED** [review]
